### PR TITLE
Decode nil into pointers for blank fields

### PR DESCRIPTION
**Pull request by andrewmostello**
The encoder encodes `nil` values to blank, but later decoding those values will result in an error unless `omitempty` is used. Adjust decoding to interpret blank values as `nil` for pointer fields, like the encoder.

### Description
A pointer may be used for a struct field when you want both `nil` and the zero value of a field to have meaning. For example, in
```
type UserInput struct {
  Agree *bool `csv:"bool"`
  Other string `csv:"other`
}
```
Agree's `true` and `false` values mean the user has input agreement, while `nil` may mean agreement has not yet been provided.

As the decoder is currently written, blank values are not allowed in decoding unless `omitempty` is specified on a field, in which case the zero values aren't marshaled. 

For example:
```
type UserInput struct {
	Agree *bool  `csv:"bool,omitempty"`
	Other string `csv:"other,omitempty"`
}

func main() {

	b := false

	exp := &UserInput{
		Agree: &b,
	}

	buf := bytes.NewBuffer(nil)
	w := csv.NewWriter(buf)
	enc := csvutil.NewEncoder(w)

	if err := enc.Encode(exp); err != nil {
		panic(err)
	}

	w.Flush()

	fmt.Println(string(buf.Bytes()))

	tst := &UserInput{}

	r := csv.NewReader(buf)
	dec, err := csvutil.NewDecoder(r)
	if err != nil {
		panic(err)
	}

	if err := dec.Decode(tst); err != nil {
		panic(err)
	}

	if reflect.DeepEqual(tst.Agree, exp.Agree) {
		fmt.Println("equal")
	} else {
		fmt.Println("not equal")
	}
}
```
prints:
```
bool,other
,

not equal
```
The omitted `false` value gets decoded back into `Agree` as `nil`, so it's not equal. To keep `false` as a valid value, remove `omitempty`, and you get:
```
bool,other
false,

equal
```
Now set `Agree` to `nil` instead of `&b`. That yields:
```
bool,other
,

panic: csvutil: cannot unmarshal  into Go value of type bool
```
You can't use the zero value if `omitempty` is specified, but you can't decode blanks if it's not. Also, an unmarshal/parse error makes sense for a value field, but for a pointer field the expectation would be to get `nil`.

These changes make that happen.

There is an ambiguous case when decoding to a string pointer. For consistency, that will also now remain `nil`. If the zero value is desired, you can just use a value field.

We've run into needing this change because we are using this to load data changes from delimited files, and we interpret no value in a field as "do not change," whereas `false` is interpreted as "set to false." It's no problem doing this with the JSON marshaler, because `null` and `false` are distinct, but here blank is either invalid or the same as `false`.

By the way, this is a great package! Thanks for putting this together.

### Checklist
- [x] Code compiles without errors
- [x] Added new tests for the provided functionality
- [x] All tests are passing
- [x] Updated the README and/or documentation, if necessary _(there is no section about pointer handling, so I didn't update anything)_

### CODE REVIEW

**Comment by jszwec** [comment]
Hi!

Thank you for bringing this up to my attention. Based on your example I realized that Encode with omitempty tag does not behave as intended for pointer values. 

The expected output for encoding `foo`:

```go
type Foo struct {
    Bool *bool `csv:",omitempty"`
}

...

var b bool
foo := Foo{ Bool: &b }
```

is:

```csv
Bool
false
```

Because bool is a non nil pointer. It should be an empty string only if Bool == nil

I think that If this behaved correctly your code would work. This is a bug in the encoder and I can fix it soon.

Your proposal for interpreting empty strings as nil pointers is a separate matter worth consideration. However, I need to sleep on it, we should be really careful with it. We have to consider all use cases first. For now I can tell that it may break backwards compatibility for pointer fields that implement `csvutil.Unmarshaler`. It's unlikely that someone is actually using it this way, but it's worth pointing it out.

**Comment by jszwec** [comment]
Im sorry it took me so long to respond. I wanted to focus on fixing the `omitempty` issues first.

I'm game for adding your changes. However, we need to fix some edge cases. I implemented two test cases that should **not** be failing. The current code won't behave correctly for pointer embedded fields. Could you please apply the following patch?

```diff
diff --git a/decoder_test.go b/decoder_test.go
index d84a127..1cae0db 100644
--- a/decoder_test.go
+++ b/decoder_test.go
@@ -194,6 +194,15 @@ type TypeI struct {
 	Int    int    `csv:"int,omitempty"`
 }

+type TypeK struct {
+	*TypeL
+}
+
+type TypeL struct {
+	String string
+	Int    int `csv:",omitempty"`
+}
+
 type Unmarshalers struct {
 	CSVUnmarshaler      CSVUnmarshaler      `csv:"csv"`
 	PCSVUnmarshaler     *CSVUnmarshaler     `csv:"pcsv"`
@@ -1017,6 +1026,22 @@ string,"{""key"":""value""}"
 			out:  &struct{ Invalid interface{} }{Invalid: &InvalidType{}},
 			err:  &UnsupportedTypeError{Type: reflect.TypeOf(InvalidType{})},
 		},
+		{
+			desc: "fails on blank non float string with ptr embedded",
+			in:   "string,float\n,",
+			out:  &TypeC{},
+			err:  &UnmarshalTypeError{Type: reflect.TypeOf(float64(0)), Value: ""},
+		},
+		{
+			desc: "blank values on embedded pointers",
+			in:   "String,Int\n,",
+			out:  &TypeK{},
+			expected: &TypeK{
+				&TypeL{String: "", Int: 0},
+			},
+			expectedRecord: []string{"", ""},
+			header:         []string{"String", "Int"},
+		},
 		{
 			desc: "blank values on pointers decode to nil",
 			in: "int,pint,int8,pint8,int16,pint16,int32,pint32,int64,pint64,uint," +
```

**Comment by jszwec** [comment]
@andrewmostello are you going to tackle this? I cannot merge it because the current implementation is buggy.

**Comment by andrewmostello** [comment]
will do - sorry I dropped off the grid here before but I'll take a look soon and get this turned around.
